File: java/src/main/java/leetcode/L0002AddTwoNumbers.java
package leetcode;

public class L0002AddTwoNumbers {

  public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    ListNode head = new ListNode(0);
    ListNode p1 = l1, p2 = l2, p0 = head;
    int carry = 0;
    while (carry != 0 || p1 != null || p2 != null) {
      int v1 = p1 != null ? p1.val : 0;
      int v2 = p2 != null ? p2.val : 0;

      // Calculate node value.
      int v0 = v1 + v2 + carry;
      carry = v0 / 10;
      v0 = v0 % 10;
      p0.next = new ListNode(v0);

      // Move to next node.
      p0 = p0.next;
      if (p1 != null) p1 = p1.next;
      if (p2 != null) p2 = p2.next;
    }

    return head.next;
  }

}

File: python3/l0002_add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        head = ListNode(None)
        r, h1, h2 = head, l1, l2
        carry = 0
        while h1 or h2 or carry:
            v1 = h1.val if h1 else 0
            v2 = h2.val if h2 else 0
            v = v1 + v2 + carry
            carry = v // 10
            r.next = ListNode(v % 10)
            r = r.next
            h1 = h1.next if h1 else None
            h2 = h2.next if h2 else None
        
        return head.next
